File: internal/model/setting/setting.go
package model

import "time"

const (
	EIGHT_HOUR_EXPIRY string = "8_hours"
	ONE_MONTH_EXPIRY  string = "1_month"
	NEVER_EXPIRY      string = "never"
)

// SystemSetting 定义系统设置实体
type SystemSetting struct {
	SiteTitle     string `json:"site_title"`     // 站点标题
	ServerName    string `json:"server_name"`    // 服务器名称
	ServerURL     string `json:"server_url"`     // 服务器地址
	AllowRegister bool   `json:"allow_register"` // 是否允许注册'
	ICPNumber     string `json:"ICP_number"`     // 备案号
	MetingAPI     string `json:"meting_api"`     // Meting API 地址
	CustomCSS     string `json:"custom_css"`     // 自定义 CSS
	CustomJS      string `json:"custom_js"`      // 自定义 JS
}

// CommentSetting 定义评论设置实体
type CommentSetting struct {
	EnableComment bool   `json:"enable_comment"` // 是否启用评论
	Provider      string `json:"provider"`       // 评论提供者
	CommentAPI    string `json:"comment_api"`    // 评论 API 地址
}

// S3Setting 定义 S3 存储设置实体
type S3Setting struct {
	Enable     bool   `json:"enable"`      // 是否启用 S3 存储
	Provider   string `json:"provider"`    // S3 服务提供商，例如 "aws", "aliyun", "minio", "other"
	Endpoint   string `json:"endpoint"`    // S3 端点
	AccessKey  string `json:"access_key"`  // 访问密钥 ID
	SecretKey  string `json:"secret_key"`  // 秘密访问密钥
	BucketName string `json:"bucket_name"` // 存储桶名称
	Region     string `json:"region"`      // 区域
	UseSSL     bool   `json:"use_ssl"`     // 是否使用 SSL
	CDNURL     string `json:"cdn_url"`     // CDN 加速域名（可选，没有就走 Endpoint）
	PathPrefix string `json:"path_prefix"` // 存储路径前缀，例如 "uploads/"，方便隔离目录
	PublicRead bool   `json:"public_read"` // 上传时是否默认设置对象为 public-read
}

// OAuth2Setting 定义 OAuth2 配置结构体
type OAuth2Setting struct {
	Enable       bool     `json:"enable"`        // 是否启用 OAuth2 登录
	Provider     string   `json:"provider"`      // OAuth2 提供商
	ClientID     string   `json:"client_id"`     // OAuth2 Client ID
	ClientSecret string   `json:"client_secret"` // OAuth2 Client Secret
	RedirectURI  string   `json:"redirect_uri"`  // OAuth2 重定向 URI
	Scopes       []string `json:"scopes"`        // OAuth2 请求的权限范围
	AuthURL      string   `json:"auth_url"`      // OAuth2 授权 URL
	TokenURL     string   `json:"token_url"`     // OAuth2 令牌 URL
	UserInfoURL  string   `json:"user_info_url"` // OAuth2 用户信息 URL
}

// AccessTokenSetting 定义访问令牌设置实体
type AccessTokenSetting struct {
	ID        int        `json:"id"`         // 访问令牌 ID
	UserID    uint       `json:"user_id"`    // 创建该访问令牌的用户 ID
	Token     string     `json:"token"`      // 访问令牌
	Name      string     `json:"name"`       // 访问令牌名称
	Expiry    *time.Time `json:"expiry"`     // 指针类型，NULL 表示永不过期
	CreatedAt time.Time  `json:"created_at"` // 访问令牌创建时间，Unix 时间戳格式
}

// FediverseSetting 定义联邦网络设置实体
type FediverseSetting struct {
	Enable    bool   `json:"enable"`     // 是否启用联邦网络功能
	ServerURL string `json:"server_url"` // 服务器 URL
}

type BackupSchedule struct {
	Enable         bool   `json:"enable"`          // 是否启用备份计划
	CronExpression string `json:"cron_expression"` // 备份计划的 Cron 表达式
}

File: web/src/stores/setting.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import {
  fetchGetSettings,
  fetchGetStatus,
  fetchGetCommentSettings,
  fetchGetS3Settings,
  fetchGetOAuth2Settings,
  fetchGetAllWebhooks,
  fetchListAccessTokens,
  fetchGetFediverseSettings,
  fetchGetBackupScheduleSetting,
  fetchHelloEch0,
} from '@/service/api'
import { localStg } from '@/utils/storage'
import { theToast } from '@/utils/toast'
import router from '@/router'
import { CommentProvider, S3Provider, OAuth2Provider } from '@/enums/enums'

export const useSettingStore = defineStore('settingStore', () => {
  /**
   * State
   */
  const isSystemReady = ref<boolean>(false)
  const SystemSetting = ref<App.Api.Setting.SystemSetting>({
    site_title: import.meta.env.VITE_APP_TITLE,
    server_name: import.meta.env.VITE_APP_NAME,
    server_url: '',
    allow_register: true,
    ICP_number: '',
    meting_api: '',
    custom_css: '',
    custom_js: '',
  })
  const CommentSetting = ref<App.Api.Setting.CommentSetting>({
    enable_comment: false,
    provider: CommentProvider.TWIKOO,
    comment_api: '',
  })
  const S3Setting = ref<App.Api.Setting.S3Setting>({
    enable: false,
    provider: S3Provider.AWS,
    endpoint: '',
    access_key: '',
    secret_key: '',
    bucket_name: '',
    region: '',
    use_ssl: false,
    cdn_url: '',
    path_prefix: '',
    public_read: true,
  })
  const OAuth2Setting = ref<App.Api.Setting.OAuth2Setting>({
    enable: false,
    provider: OAuth2Provider.GITHUB,
    client_id: '',
    client_secret: '',
    redirect_uri: '',
    scopes: [],
    auth_url: '',
    token_url: '',
    user_info_url: '',
  })
  const Webhooks = ref<App.Api.Setting.Webhook[]>([])
  const AccessTokens = ref<App.Api.Setting.AccessToken[]>([])
  const FediverseSetting = ref<App.Api.Setting.FediverseSetting>({
    enable: false,
    server_url: '',
  })
  const BackupSchedule = ref<App.Api.Setting.BackupSchedule>({
    enable: false,
    cron_expression: '0 2 * * 0',
  })
  const hello = ref<App.Api.Ech0.HelloEch0>()
  const loading = ref<boolean>(true)

  /**
   * Actions
   */
  const getSystemReady = async () => {
    // 检查localStorage中是否有系统状态
    const systemStatus = localStg.getItem<boolean>('systemStatus')
    if (systemStatus !== null) {
      // 如果有，直接使用localStorage中的值
      isSystemReady.value = systemStatus
    } else {
      // 如果没有，默认设置为false
      isSystemReady.value = false
    }

    // 检查系统是否准备好
    if (!isSystemReady.value) {
      // 如果系统未准备好，调用接口获取系统状态
      const res = await fetchGetStatus()
      if (res.code === 666) {
        isSystemReady.value = false
        theToast.info(res.msg)
        // 跳转到注册页面
        router.push({ name: 'auth' })
      } else {
        isSystemReady.value = true
        console.log('系统已准备好')
      }

      // 保存系统状态到localStorage
      localStg.setItem('systemStatus', isSystemReady.value)
    }
  }

  const getSystemSetting = async () => {
    await fetchGetSettings().then((res) => {
      if (res.code === 1) {
        SystemSetting.value = res.data
        loading.value = false
      }
    })
  }

  const getCommentSetting = async () => {
    fetchGetCommentSettings().then((res) => {
      if (res.code === 1) {
        CommentSetting.value = res.data
      }
    })
  }

  const getS3Setting = async () => {
    fetchGetS3Settings().then((res) => {
      if (res.code === 1) {
        S3Setting.value = res.data
      }
    })
  }

  const getOAuth2Setting = async () => {
    const res = await fetchGetOAuth2Settings()
    if (res.code === 1) {
      OAuth2Setting.value = res.data
    }
    return OAuth2Setting.value
  }

  const getAllWebhooks = async () => {
    const res = await fetchGetAllWebhooks()
    if (res.code === 1) {
      if (res.data) {
        Webhooks.value = res.data
      } else {
        Webhooks.value = []
      }
    }
  }

  const getAllAccessTokens = async () => {
    const res = await fetchListAccessTokens()
    if (res.code === 1) {
      if (res.data) {
        AccessTokens.value = res.data
      } else {
        AccessTokens.value = []
      }
    }
  }

  const getFediverseSetting = async () => {
    const res = await fetchGetFediverseSettings()
    if (res.code === 1) {
      FediverseSetting.value = res.data
    }
  }

  const getBackupSchedule = async () => {
    const res = await fetchGetBackupScheduleSetting()
    if (res.code === 1) {
      BackupSchedule.value = res.data
    }
  }

  const getHelloEch0 = async () => {
    const res = await fetchHelloEch0()
    if (res.code === 1) {
      hello.value = res.data
    }
  }

  const setSystemReady = (status: boolean) => {
    isSystemReady.value = status
  }

  const init = () => {
    if (!isSystemReady.value) {
      getSystemReady()
    }
    getSystemSetting()
    getCommentSetting()
    getS3Setting()
    getHelloEch0()
  }

  return {
    isSystemReady,
    SystemSetting,
    CommentSetting,
    S3Setting,
    OAuth2Setting,
    Webhooks,
    AccessTokens,
    FediverseSetting,
    BackupSchedule,
    hello,
    loading,

    getAllAccessTokens,
    getSystemReady,
    getSystemSetting,
    getCommentSetting,
    getS3Setting,
    getOAuth2Setting,
    getAllWebhooks,
    getFediverseSetting,
    setSystemReady,
    getHelloEch0,
    getBackupSchedule,
    init,
  }
})
